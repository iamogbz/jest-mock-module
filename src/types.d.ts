import { ValueOf } from "jest-mock-props/lib/types";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type AnyRecord = Record<string, any>;
export type AnyObject = object & AnyRecord;
export type Mock<O> = {
    [K in string]: Mock<ValueOf<O>>;
};

declare global {
    namespace jest {
        function spy(moduleName: string): void;
    }
}
