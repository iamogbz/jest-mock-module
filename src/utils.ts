import { Entries, MapFn, Mapped } from "../typings/globals";

/**
 * Only maps on the string properties using Object.entries
 * @param o the object to map on
 * @param callbackfn the function to be called with object entries returning new value
 */
export function mapObject<ObjectType, MappedObjectType extends ObjectType>(
    o: ObjectType,
    callbackfn: MapFn<ObjectType, MappedObjectType>,
): Mapped<ObjectType, MappedObjectType> {
    const entries = Object.entries(o) as Entries<ObjectType>;
    const newValues = entries.map(callbackfn);
    const newObject = {} as Mapped<ObjectType, MappedObjectType>;
    entries.forEach(([key], i) => (newObject[key] = newValues[i]));
    return newObject;
}
