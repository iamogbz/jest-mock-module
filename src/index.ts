import { ExtendJest, Mock, Optional, SpyOnModule } from "../typings/globals";
import { mapObject } from "./utils";

function maybeSpyOnProp<T>(object: T, propName: keyof T) {
    try {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        return require("jest-mock-props").spyOnProp(object, propName);
    } catch (e) {
        // eslint-disable-next-line no-console
        console.warn(e.message);
        return object[propName];
    }
}

function spyOn<T>(object: T, propName: keyof T) {
    const propValue = object[propName];
    const propType = typeof propValue;
    if (propType === "function") {
        // @ts-expect-error Jest does not play nice
        return jest.spyOn(object, propName);
    }
    if (propType === "object") {
        return spyOnObject(propValue);
    }
    return maybeSpyOnProp(object, propName);
}

export function spyOnObject<T>(o?: T): Optional<Mock<T>> {
    if (o === undefined || o === null) return o;
    return mapObject(o, ([k]) => spyOn(o, k));
}

export const spy: SpyOnModule = (moduleName) => {
    const actualModule = jest.requireActual(moduleName);
    jest.mock(moduleName, () => spyOnObject(actualModule));
};

export const extend: ExtendJest = (jestInstance) => {
    try {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        require("jest-mock-props").extend(jestInstance);
    } catch (e) {
        // eslint-disable-next-line no-console
        console.info(
            "Install package 'jest-mock-props' to " +
                "spy on non function module properties.",
        );
    }
    Object.assign(jestInstance, { spy });
};

export * from "../typings/globals";
