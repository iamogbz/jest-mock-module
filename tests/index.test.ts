import "jest-mock-props";
import { extend, spyOnObject } from "src/index";
extend(jest);
afterEach(jest.clearAllMocks);

it("spies on methods called", () => {
    jest.spy("tests/mocks/hello");
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { say, shout } = require("./mocks/hello");
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const sayHello = require("./mocks/index").default;

    expect(sayHello("You")).toEqual("HELLO YOU!");
    expect(say).toHaveBeenCalledTimes(1);
    expect(shout).toHaveBeenCalledTimes(1);
    ((say as unknown) as jest.SpyInstance).mockImplementationOnce(
        () => "something",
    );
    expect(sayHello("Who")).toEqual("SOMETHING!");
    ((shout as unknown) as jest.SpyInstance).mockImplementationOnce(
        (s: string) => s.toLocaleLowerCase(),
    );
    expect(sayHello("Gru")).toEqual("hello gru");
});

it("spies on object nested properties", () => {
    const obj = {
        prop1: 1,
        prop2: "two",
        propFn(something: string) {
            return `say ${something}`;
        },
        nested: {
            propA: null,
            propB: true,
        },
    };
    const spied = spyOnObject(jest, obj);
    if (!spied) fail("Expect spied to be defined and not null");
    expect(obj).toMatchSnapshot();
    expect(jest.isMockProp(obj, "prop1")).toBe(true);
    expect(jest.isMockProp(obj, "prop2")).toBe(true);
    expect(jest.isMockFunction(obj.propFn)).toBe(true);
    expect(jest.isMockProp(obj, "nested")).toBe(false);
    expect(jest.isMockProp(obj.nested, "propA")).toBe(true);
    expect(jest.isMockProp(obj.nested, "propB")).toBe(true);
});
