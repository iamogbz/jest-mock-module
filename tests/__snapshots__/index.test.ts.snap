// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spys on object nested properties 1`] = `
Object {
  "nested": Object {
    "propA": null,
    "propB": true,
  },
  "prop1": 1,
  "prop2": "two",
  "propFn": [MockFunction],
}
`;
